### PR TITLE
preserve expected-type explanation after "let () = ... in"

### LINKED ISSUE #10106: Conditionals with a missing else-branch produce an unclear type error with let-expressions
If a conditional with a missing `else` branch contains only a simple boolean expression, then the compiler tells you what's wrong: missing `else` means implicit `unit`.

```ocaml
# if false then true ;;
Line 1, characters 14-18:
Error: This variant expression is expected to have type unit
         because it is in the result of a conditional with no else branch
       The constructor true does not belong to type unit

```
However, if you put a `let`-expression there, the compiler simply tells you that the expression is expected to have type `unit`, but no longer tells you the reason.

```ocaml
# 
if false then
  let () = print_endline "false is the new true" in
  true 
;;

Line 1, characters 64-68:
Error: This variant expression is expected to have type unit
       The constructor true does not belong to type unit

```

If would be nice if it could give a good error message in that situation too.

### PR DESCRIPTION
**Pull request by gasche**
fixes #10106

### CODE REVIEW

File: testsuite/tests/typing-core-bugs/type_expected_explanation.ml [hunk 1]
```diff
@@ -187,3 +187,17 @@ Error: This expression has type int but an expression was expected of type
...
+|}];;
+
+[%%expect{|
+Line 1, characters 35-39:
+1 | if false then (match () with () -> true);;
+                                       ^^^^
+Error: This variant expression is expected to have type unit
+         because it is in the result of a conditional with no else branch
+       The constructor true does not belong to type unit
+|}]
```
Comment (trefis) [inline testsuite/tests/typing-core-bugs/type_expected_explanation.ml]: What's going on here? Why are there two expect blocks?

**Review by trefis — APPROVED** [review]
Code changes look good.

File: Changes [hunk 2]
```diff
@@ -767,6 +767,9 @@ OCaml 4.12.0
 - #10048: Fix bug with generalized local opens.
   (Leo White, review by Thomas Refis)
 
+- #10106, #10112: preserve expected-type explanation after "let () = ... in"
```
Comment (Octachron) [inline Changes]: The fix is quite broader than just `let ( ) = ... in` . Maybe

> after some instances of `let ... = ... in` and `try ... with`

?

Comment (dmbaturin) [reply Changes]: Perhaps something like: Preserve the type error explanation in conditionals with a non-unit let-expression in the "if" branch and a missing "else" branch.

File: testsuite/tests/typing-core-bugs/type_expected_explanation.ml (hunk 1 above)
Comment (gasche) [reply testsuite/tests/typing-core-bugs/type_expected_explanation.ml]: Good catch, thanks; fixed.

File: Changes (hunk 2 above)
Comment (gasche) [reply Changes]: I rephrased the Changes entry, thanks.

**Comment by gasche** [comment]
Thanks for the reviews, I rebased the PR to take review comments into account. My impression is that this should be good to go if the CI agrees.

@Octachron you commented on the phrasing of the Changes entry, which I take as a low-key way to signal that you are totally fine with the entry being placed in the 4.12 section, and approve of cherry-picking in the release branch. (My reasoning is that it fixes a (small) bug and the change is rather safe.)

**Comment by Octachron** [comment]
I did not notice that the Change entry was in the 4.12 section; but yes fixing the propagation of explanation for a potential errors seems safe enough for 4.12.0 .

**Comment by gasche** [comment]
Cherry-picked in 4.12 as 671284bd552c271985d08fdb7a7b69ee29560cee.

**Pull request merged**